### PR TITLE
Added a driver for the OpenTitan I2C block

### PR DESCRIPTION
**Pull request by HU90m**
Implemented a driver for the OpenTitan I2C block. This was tested the demo in https://github.com/lowRISC/sonata-software/pull/4.

*Also implemented `isprint()` in the standard library.*

### CODE REVIEW

File: sdk/include/platform/sunburst/platform-i2c.hh [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+			while (!fmt_empty()) {}
+			if (interrupt_asserted(OpenTitanI2cIntr::Nak))
+			{
+				clear_interrupt(OpenTitanI2cIntr::Nak);
+				return false;
+			}
+			uint32_t bytesRemaining = NumBytes - idx;
+			bool     lastChunk      = UINT8_MAX >= bytesRemaining;
+			uint8_t  chunkSize =
+              lastChunk ? static_cast<uint8_t>(bytesRemaining) : UINT8_MAX;
```
Comment (HU90m) [inline sdk/include/platform/sunburst/platform-i2c.hh:312]: This is what `clang-format` decided was best :frowning_face:

**Review by marnovandermaas — APPROVED** [review]
Looks reasonable to me although I haven't done any testing yet.

**Comment by HU90m** [comment]
@GregAC would it be possible to change the repository default to 'rebase and merge'.

File: sdk/include/platform/sunburst/platform-i2c.hh [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+
+/**
+ * Performs a 32-bit integer unsigned division, rounding up. The bottom
+ * 16 bits of the result are then returned.
+ *
+ * As usual, a divisor of 0 is still Undefined Behavior.
+ */
+static constexpr uint16_t round_up_divide(const uint32_t A, const uint32_t B)
+{
+	const uint32_t Result = ((A - 1) / B) + 1;
```
Comment (nbdd0121) [inline sdk/include/platform/sunburst/platform-i2c.hh:27]: This doesn't work with A = 0.

File: sdk/include/platform/sunburst/platform-i2c.hh [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+ */
+static const uint16_t SpcThigh[]  = {4000u, 600u, 260u};
+static const uint16_t SpcTlow[]   = {4700u, 1300u, 150u};
+static const uint16_t SpcThdSta[] = {4000u, 600u, 260u};
+static const uint16_t SpcTsuSta[] = {4700u, 600u, 260u};
+static const uint16_t SpcThdDat[] = {5000u, 1u, 1u};
+static const uint16_t SpcTsuDat[] = {250u, 100u, 50u};
+static const uint16_t SpcTBuf[]   = {4700u, 1300u, 500u};
+static const uint16_t SpcTsuSto[] = {4000u, 600u, 260u};
+
```
Comment (nbdd0121) [inline sdk/include/platform/sunburst/platform-i2c.hh:18]: Wouldn't this pollute the global namespace?

File: sdk/include/platform/sunburst/platform-i2c.hh [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+	/**
+	 * Set the I2C timing parameters appropriately for the given bit rate.
+	 * Distilled from:
+	 * https://opentitan.org/book/hw/ip/i2c/doc/programmers_guide.html
+	 */
+	void set_speed(const uint32_t SpeedKhz) volatile
+	{
+		// We must round up the system clock frequency to lengthen intervals.
+		constexpr uint32_t SysclkKhz = (I2cDevClockHz + 999) / 1000;
```
Comment (nbdd0121) [inline sdk/include/platform/sunburst/platform-i2c.hh:235]: Use `round_up_divide` here?

File: sdk/include/ctype.h [hunk 5]
```diff
@@ -3,6 +3,17 @@
...
+static inline int isprint(int c)
+{
+	switch (c)
+	{
+		default:
+			return 0;
+		case '\x20' ... '\x7e':
+			return 1;
+	}
+}
```
Comment (nbdd0121) [inline sdk/include/ctype.h:15]: This could just be a one-liner.

File: sdk/include/platform/sunburst/platform-i2c.hh (hunk 4 above)
Comment (HU90m) [reply sdk/include/platform/sunburst/platform-i2c.hh:235]: This is a 32-bit value, round_up_divide truncates to a 16-bit value

**Pull request merged**